service/storage_service.cc

```diff
@@ -6880,6 +6878,18 @@ future<> storage_service::move_tablet(table_id table, dht::token token, locator:
             co_return;
         }
 
+        if (locator::contains(tinfo.replicas, dst.host)) {
+            throw std::runtime_error(format("Tablet {} has replica on {}", gid, dst.host));
+        }
+        auto src_dc_rack = get_dc_rack_for(host2ip(src.host)).value();
+        auto dst_dc_rack = get_dc_rack_for(host2ip(dst.host)).value();
+        if (src_dc_rack.dc != dst_dc_rack.dc) {
+            throw std::runtime_error(format("Attempted to move tablet {} between DCs ({} and {})", gid, src_dc_rack.dc, dst_dc_rack.dc));
+        }
+        if (src_dc_rack.rack != dst_dc_rack.rack) {
+            throw std::runtime_error(format("Attempted to move tablet {} between racks ({} and {})", gid, src_dc_rack.rack, dst_dc_rack.rack));
+        }
```

Comment:
I think the rack check is too strict. It's ok to move across racks as long as it doesn't reduce HA, meaning number of replicas per rack is not greater than before.